### ISSUE TITLE
Berry curvature $\Omega_{n,z}$(k)

### ISSUE THREAD

**Issue by Guangzongxing** [wannier-berri/wannier-berri#225]
Dear Prof.  Stepan Tsirkin,

Thank you for the useful code. I am a new user to WannierBerri and want to evaluate the k-vecotor decomposed anomalous Nernst conductivity at different temperature, which is a sum of the  z-component of berry curvature  $\Omega_{n,z}$ (k) times a temperature dependent function, where n being the band index.   Can I get the berry curvature $\Omega_{n,z}$ at each n? If I understand correctly, the obtained tabulated value $\Omega_{z}$(k)  is the the sum of all the Wannier band n. I want to evaluate the quantity $\Omega_{n,z}\times{}g(T)$ first and then do the summation for all n at each k.  Thank you for your help.

**Comment by Guangzongxing** [comment]
I am really sorry for the very simple question. I think the iband parameter controls the band index. I will have a try to obtain $\Omega_{n,z}$ with iband for each n. Thank you again.

**Comment by stepan-tsirkin** [comment]
@Guangzongxing , yes `iband`  parameter of [TabulatorAll](https://wannier-berri.org/docs/calculators.html#wannierberri.calculators.TabulatorAll) controls which bands to plot.

However, tabulating Berry curvature and storing on disk will be a very inefficient way of computing ANC. A better way is to integrate Berry curvature weighted by  a delta-function $\delta(\varepsilon_{nk}-\mu)$ (once for all, using tetrahedron method for a dense grid of $\mu$) Then applying any distribution function (for any temperature of chemical potential) will be just a one-dimensional integral, which you can easily perform with your script.

See here for more info: https://wannier-berri.org/tutorials/DIY/tutorial-wb-DIY.html
in fact, AHC is not included in WanierBerri  for the only one reason - to have the oppeorunity to provide such a simple tutorial. 

Please do not hesitate to ask questions, if you have.

**Comment by Guangzongxing** [comment]
Dear Pr. @stepan-tsirkin, thank you for your kind reply.  I am still a little bit confused about the method you describe to calculated the k-momentum decomposed ANC. As we know, the ANH is 
### $\alpha_{xy} = -\frac{1}{T}\frac{e}{\hbar}\sum_n\int \frac{dk^3}{(2\pi)^3}\Omega_{xy}^{n}[(\epsilon_{n}-\epsilon_{F})f_{n}+k_BT\ln(1+\exp(\frac{\epsilon_{n}-\epsilon_{F}}{-k_BT}))]$
where $f_n$ and $\epsilon_{F}$ and $\epsilon_n$ are the Fermi distribution and the Fermi level and the eigenenergy with the band index n. Now, I want to make a same k-decomposed ANC plot as the attached plot ( Fig. (b) for ANC) from the  PRB paper [https://journals.aps.org/prb/abstract/10.1103/PhysRevB.98.241106](url). If I understand correctly what I need to compute the following 
#### $\sum_n\Omega_{xy}^{n}s(\epsilon_n)$. with $s(\epsilon_n) = (\epsilon_{n}-\epsilon_{F})f_{n}+k_BT\ln(1+\exp(\frac{\epsilon_{n}-\epsilon_{F}}{-k_BT}))$
Then, what I really need to compute then is the $\Omega_{xy}^{n}$ at each $k$. It is obvious that the $\Omega_{xy}^{n}$ can be obtained by Wannierberri with the parameter iband. 

With respect to the second method you describe, my question is

#### About integrating the Berry curvature with a delta function, is the integration of $\Omega_{xy}^{n}\delta(\epsilon_{n,k}-\mu))$ with the dense $\mu$  grid variable to be integrated?  

#### What will obtain after the integration?

#### Do I have to sum all the $\int_{\mu}\Omega_{xy}^{n}\delta(\epsilon_{n,k}-\mu)$ with the band index n  first and then times $s(\epsilon_n)$ or times $s(\epsilon_n)$ first and then do

 the summation over n ?

I appreciate your help.

<img width="473" alt="ANC" src="https://user-images.githubusercontent.com/104555911/215024805-a48f3f9e-f496-484e-8952-e64b4aea8bfa.png">

**Comment by stepan-tsirkin** [comment]
Dear @Guangzongxing ,
i am sorry, I did not read your question carefully, I did not get that you need the k-resolved quantity. For that you really need to tabulate.  (The method I described is only good for k-integrated quantity)

I noted that tabulating is a bit less efficient, in particular because of writing 'frmsf' formatted files. With the following [branch](https://github.com/wannier-berri/wannier-berri/pull/226) (to be merged soon)  you can disable writing the frmsf files, and work with arrays directly, using something like this:

```
result = wberri.run(system,
            grid=grid,
            calculators = {
                 "tabulate":wberri.calculators.TabulatorAll({
                            "Energy":wberri.calculators.tabulate.Energy(),
                            "berry":wberri.calculators.tabulate.BerryCurvature(),
                                  }, 
                                    save_mode = "",
                                       ibands = np.arange(4,10)),
                          }, 
            parallel=parallel,
            adpt_num_iter=0,
            restart=False,
            )
t_run = time()-t0
print ("t_run = ",t_run)
grid_result = result.results["tabulate"]
Energy = grid_result.get_data(iband=5, quantity='Energy')
berry  = grid_result.get_data(iband=5, quantity='berry',component='z')
print(berry.shape,Energy.shape)

```
